File: dart_sample/lib/demos/cupertino/picker_demo_page.dart
import 'package:cupertino_ui/cupertino_ui.dart';
import 'package:material_ui/material_ui.dart';

class CupertinoPickerDemoPage extends StatefulWidget {
  const CupertinoPickerDemoPage({super.key});

  @override
  State<CupertinoPickerDemoPage> createState() =>
      _CupertinoPickerDemoPageState();
}

class _CupertinoPickerDemoPageState extends State<CupertinoPickerDemoPage> {
  static const List<String> _fruits = <String>[
    'Apple',
    'Banana',
    'Cherry',
    'Pear',
  ];
  static const List<String> _sizes = <String>[
    'Small',
    'Medium',
    'Large',
    'Extra large',
  ];

  final FixedExtentScrollController _fruitController =
      FixedExtentScrollController(initialItem: 1);
  final FixedExtentScrollController _sizeController =
      FixedExtentScrollController(initialItem: 1);
  int _fruitIndex = 1;
  int _sizeIndex = 1;

  @override
  Widget build(BuildContext context) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 12,
      children: <Widget>[
        const Text(
          'Cupertino picker',
          style: TextStyle(fontSize: 20, color: Colors.black),
        ),
        const Text(
          'Scroll or tap a row. The left wheel loops; the right wheel uses '
          'the lazy builder API.',
          style: TextStyle(fontSize: 14, color: Colors.black54),
        ),
        Text(
          'Fruit: ${_fruits[_fruitIndex]} · Size: ${_sizes[_sizeIndex]}',
          style: const TextStyle(fontSize: 13, color: Colors.blueGrey),
        ),
        Row(
          spacing: 16,
          children: <Widget>[
            Expanded(
              child: _buildWheelColumn(
                'Looping list',
                CupertinoPicker(
                  itemExtent: 40,
                  onSelectedItemChanged: _selectFruit,
                  scrollController: _fruitController,
                  looping: true,
                  children: _fruits
                      .map(
                        (String fruit) =>
                            Text(fruit, textAlign: TextAlign.center),
                      )
                      .toList(),
                ),
              ),
            ),
            Expanded(
              child: _buildWheelColumn(
                'Builder + magnifier',
                CupertinoPicker.builder(
                  itemExtent: 40,
                  onSelectedItemChanged: _selectSize,
                  itemBuilder: (BuildContext context, int index) {
                    return Text(_sizes[index], textAlign: TextAlign.center);
                  },
                  selectionOverlay:
                      const CupertinoPickerDefaultSelectionOverlay(
                        capStartEdge: false,
                      ),
                  childCount: _sizes.length,
                  useMagnifier: true,
                  magnification: 1.12,
                  scrollController: _sizeController,
                ),
              ),
            ),
          ],
        ),
      ],
    );
  }

  @override
  void dispose() {
    _fruitController.dispose();
    _sizeController.dispose();
    super.dispose();
  }

  Widget _buildWheelColumn(String label, Widget picker) {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.stretch,
      spacing: 6,
      children: <Widget>[
        Text(
          label,
          textAlign: TextAlign.center,
          style: const TextStyle(fontSize: 13, color: Color(0xFF37474F)),
        ),
        SizedBox(height: 180, child: picker),
      ],
    );
  }

  void _selectFruit(int index) {
    setState(() {
      _fruitIndex =
          ((index % _fruits.length) + _fruits.length) % _fruits.length;
    });
  }

  void _selectSize(int index) {
    setState(() {
      _sizeIndex = index;
    });
  }
}

File: src/Sample/Plumix.Sample/Demos/Cupertino/PickerDemoPage.cs
using System;
using System.Collections.Generic;
using Avalonia.Media;
using Plumix.Cupertino;
using Plumix.Rendering;
using Plumix.UI;
using Plumix.Widgets;

// Dart parity source (reference): dart_sample/lib/demos/cupertino/picker_demo_page.dart
// (exact sample parity)

namespace Plumix;

public sealed class CupertinoPickerDemoPage : StatefulWidget
{
    public override State CreateState() => new CupertinoPickerDemoPageState();
}

internal sealed class CupertinoPickerDemoPageState : State
{
    private static readonly IReadOnlyList<string> Fruits = ["Apple", "Banana", "Cherry", "Pear"];
    private static readonly IReadOnlyList<string> Sizes = ["Small", "Medium", "Large", "Extra large"];

    private readonly FixedExtentScrollController _fruitController = new(initialItem: 1);
    private readonly FixedExtentScrollController _sizeController = new(initialItem: 1);
    private int _fruitIndex = 1;
    private int _sizeIndex = 1;

    public override Widget Build(BuildContext context)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 12.0,
            children:
            [
                new Text("Cupertino picker", fontSize: 20.0, color: Colors.Black),
                new Text(
                    "Scroll or tap a row. The left wheel loops; the right wheel uses the lazy builder API.",
                    fontSize: 14.0,
                    color: Color.Parse("#8A000000")),
                new Text(
                    $"Fruit: {Fruits[_fruitIndex]} · Size: {Sizes[_sizeIndex]}",
                    fontSize: 13.0,
                    color: Color.Parse("#FF607D8B")),
                new Row(
                    spacing: 16.0,
                    children:
                    [
                        new Expanded(BuildWheelColumn(
                            "Looping list",
                            new CupertinoPicker(
                                itemExtent: 40.0,
                                onSelectedItemChanged: SelectFruit,
                                children: BuildFruitChildren(),
                                scrollController: _fruitController,
                                looping: true))),
                        new Expanded(BuildWheelColumn(
                            "Builder + magnifier",
                            CupertinoPicker.Builder(
                                itemExtent: 40.0,
                                onSelectedItemChanged: SelectSize,
                                itemBuilder: (_, index) => new Text(Sizes[index], textAlign: TextAlign.Center),
                                selectionOverlay: new CupertinoPickerDefaultSelectionOverlay(
                                    capStartEdge: false),
                                childCount: Sizes.Count,
                                useMagnifier: true,
                                magnification: 1.12,
                                scrollController: _sizeController))),
                    ]),
            ]);
    }

    public override void Dispose()
    {
        _fruitController.Dispose();
        _sizeController.Dispose();
        base.Dispose();
    }

    private static IReadOnlyList<Widget> BuildFruitChildren()
    {
        var children = new List<Widget>(Fruits.Count);
        foreach (string fruit in Fruits)
        {
            children.Add(new Text(fruit, textAlign: TextAlign.Center));
        }

        return children;
    }

    private static Widget BuildWheelColumn(string label, Widget picker)
    {
        return new Column(
            crossAxisAlignment: CrossAxisAlignment.Stretch,
            spacing: 6.0,
            children:
            [
                new Text(label, fontSize: 13.0, color: Color.Parse("#FF37474F"), textAlign: TextAlign.Center),
                new SizedBox(height: 180.0, child: picker),
            ]);
    }

    private void SelectFruit(int index)
    {
        SetState(() => _fruitIndex = ((index % Fruits.Count) + Fruits.Count) % Fruits.Count);
    }

    private void SelectSize(int index)
    {
        SetState(() => _sizeIndex = index);
    }
}
